finetuner/__init__.py

```diff
@@ -28,8 +29,13 @@ def fit(
     epochs: int = 10,
     batch_size: int = 256,
     loss: Union[str, 'AnyDNN'] = 'SiameseLoss',
-    optimizer: Optional['AnyOptimizer'] = None,
```

Comment:
why u can't pass an optimizer object?

Reply:
In python we move the model to a GPU if needed. However, if this is done after optimizer is initialized, this will break the optimizer

> If you need to move a model to GPU via .cuda(), please do so before constructing optimizers for it. Parameters of a model after .cuda() will be different objects with those before the call.
> In general, you should make sure that optimized parameters live in consistent locations when optimizers are constructed and used.

Reply:
The problem is that u are moving `fit` arguments to `attributes` of the `objects`. What a `Tuner` is should be determined by the model and the `loss` perhaps. `Optimizers` and `learning_rates` should be different call arguments possible with the same `object`.

Reply:
Well no, the problem persists, and the factory idea is good, but they should remain call arguments and not attributes

Reply:
I have thought about this, and I have decided to go against it. What is an "attribute" of Tuner and what is not is just semantics really - I can easily argue that because this is `Tuner` and not `Model` all things crucial to tuning need to go in `__init__`. 

I have instead a very concrete argument - all the things whose state can change during training should go into `__init__`. This is due to checkpointing - if my training is interrupted in epoch 5, and I want to later start from that position, I need a way of restoring that state. There are two ways of doing this:
1. Having the `fit` method already start with initialized states. In this case, I can, right after initializing the trainer, call `TrainingCheckpoint.load(tuner, path_to_checkpoint)`. This is my preferred option, as it decouples checkpoint from the trainer.
2. Having the checkpoint and training tightly coupled, so that you do `tuner.fit(load_checkpoint=path_to_checkpoint)`. I don't really like this option.

And the way this is done in different frameworks: 
- in keras you call `compile` on a model first (equivalent of our trainer init), passing optimizer with learning rate, metrics, etc. In `fit` you only pass data and related args
- in pytorch lightning you pass all the configuration (including things like number of epochs, device) in init, and pass model with data (optimizer  is specified as part of the model definition) and checkpoint path in `fit` - here they have checkpointing tightly coupled with the training method
- in pytorch ignite you have to create all objects before calling `Engine.run` - which takes only data and number of epochs.

Based on the prefered way of implementing checkpoints, and the observation that this is indeed how it is also done in other frameworks, I would go ahead with my implementation.